File: solutions/clojure/113-basic-calculator.clj
;; Problem 113: Basic Calculator (LeetCode 224)
;; Difficulty: Hard
;; Language: Clojure
;; 
(defn calculate [s]
  (let [chars (vec (remove #(= % \space) s))]
    (letfn [(parse [i] (go i 0 1))
            (go [i result sign]
              (cond
                (>= i (count chars)) [result i]
                (= (chars i) \)) [result (inc i)]
                (= (chars i) \+) (go (inc i) result 1)
                (= (chars i) \-) (go (inc i) result -1)
                (= (chars i) \() (let [[v ni] (parse (inc i))]
                                    (go ni (+ result (* sign v)) 1))
                :else (let [[num ni] (read-num i 0)]
                        (go ni (+ result (* sign num)) 1))))
            (read-num [i acc]
              (if (and (< i (count chars))
                       (Character/isDigit (chars i)))
                (read-num (inc i)
                  (+ (* acc 10) (Character/digit (chars i) 10)))
                [acc i]))]
      (first (parse 0)))))

File: solutions/rust/113-basic-calculator.rs
/**
 * Problem 113: Basic Calculator (LeetCode 224)
 * Difficulty: Hard
 * Language: Rust
 */
fn calculate(s: &str) -> i32 {
    let chars: Vec<char> = s.chars().filter(|c| !c.is_whitespace()).collect();
    fn parse(chars: &[char], mut i: usize) -> (i32, usize) {
        let (mut result, mut sign) = (0i32, 1i32);
        while i < chars.len() && chars[i] != ')' {
            match chars[i] {
                '+' => { sign = 1; i += 1; }
                '-' => { sign = -1; i += 1; }
                '(' => {
                    let (v, ni) = parse(chars, i + 1);
                    result += sign * v; sign = 1; i = ni;
                }
                _ => {
                    let mut num = 0i32;
                    while i < chars.len() && chars[i].is_ascii_digit() {
                        num = num * 10 + (chars[i] as i32 - '0' as i32);
                        i += 1;
                    }
                    result += sign * num; sign = 1;
                }
            }
        }
        (result, i + 1)
    }
    parse(&chars, 0).0
}
